api/metrics/v1/alertmanager_enforcer.go

```diff
@@ -0,0 +1,130 @@
+package v1
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/observatorium/api/authentication"
+	"github.com/observatorium/api/httperr"
+	"github.com/prometheus/alertmanager/api/v2/models"
+	amlabels "github.com/prometheus/alertmanager/pkg/labels"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+// WithEnforceTenancyOnFilter returns a middleware that ensures that every filter has a tenant label enforced.
+func WithEnforceTenancyOnFilter(label string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		// https://github.com/prometheus-community/prom-label-proxy/
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			id, ok := authentication.GetTenantID(r.Context())
+			if !ok {
+				httperr.PrometheusAPIError(w, "error finding tenant ID", http.StatusInternalServerError)
+
+				return
+			}
+
+			matcher := &labels.Matcher{
+				Name:  label,
+				Type:  labels.MatchEqual,
+				Value: id,
+			}
+			matcherStr := matcher.String()
+
+			q := r.URL.Query()
+			filters := q["filter"]
+			modified := []string{matcherStr}
+
+			if len(filters) == 0 {
+				q.Set("filter", matcherStr)
+			} else {
+
+				for _, filter := range filters {
+					m, err := amlabels.ParseMatcher(filter)
+					if err != nil {
+						return
+					}
+					// Keep the original matcher in case of multi label values because
+					// the user might want to filter on a specific value.
+					if m.Name == label {
+						continue
+					}
```

Comment:
What's the behaviour in case we get a request with a `tenant` label, that has a different value? We just don't process it and default to tenant one?

Reply:
yup, thats the intention, otherwise they could impersonate right?

Reply:
Ack!